### PR TITLE
[RFC] vim-patch:7.4.636

### PR DESCRIPTION
**Pull request by glts**
```
Problem:    A search with end offset gets stuck at end of file. (Gary Johnson)
Solution:   When a search doesn't move the cursor repeat it with a higher
            count. (Christian Brabandt)
```

https://code.google.com/p/vim/source/detail?r=v7-4-636

This is in preparation for porting of legacy tests 44 and 99.

Original patch:

``` diff
diff --git a/src/nvim/normal.c b/src/nvim/normal.c
--- a/src/nvim/normal.c
+++ b/src/nvim/normal.c
@@ -100,7 +100,7 @@ static void nv_end __ARGS((cmdarg_T *cap
 static void    nv_dollar __ARGS((cmdarg_T *cap));
 static void    nv_search __ARGS((cmdarg_T *cap));
 static void    nv_next __ARGS((cmdarg_T *cap));
-static void    normal_search __ARGS((cmdarg_T *cap, int dir, char_u *pat, int opt));
+static int normal_search __ARGS((cmdarg_T *cap, int dir, char_u *pat, int opt));
 static void    nv_csearch __ARGS((cmdarg_T *cap));
 static void    nv_brackets __ARGS((cmdarg_T *cap));
 static void    nv_percent __ARGS((cmdarg_T *cap));
@@ -5765,7 +5765,7 @@ nv_ident(cap)
    init_history();
    add_to_history(HIST_SEARCH, buf, TRUE, NUL);
 #endif
-   normal_search(cap, cmdchar == '*' ? '/' : '?', buf, 0);
+   (void)normal_search(cap, cmdchar == '*' ? '/' : '?', buf, 0);
     }
     else
    do_cmdline_cmd(buf);
@@ -6301,7 +6301,7 @@ nv_search(cap)
    return;
     }

-    normal_search(cap, cap->cmdchar, cap->searchbuf,
+    (void)normal_search(cap, cap->cmdchar, cap->searchbuf,
                        (cap->arg ? 0 : SEARCH_MARK));
 }

@@ -6313,14 +6313,26 @@ nv_search(cap)
 nv_next(cap)
     cmdarg_T   *cap;
 {
-    normal_search(cap, 0, NULL, SEARCH_MARK | cap->arg);
+    pos_T old = curwin->w_cursor;
+    int   i = normal_search(cap, 0, NULL, SEARCH_MARK | cap->arg);
+
+    if (i == 1 && equalpos(old, curwin->w_cursor))
+    {
+   /* Avoid getting stuck on the current cursor position, which can
+    * happen when an offset is given and the cursor is on the last char
+    * in the buffer: Repeat with count + 1. */
+   cap->count1 += 1;
+   (void)normal_search(cap, 0, NULL, SEARCH_MARK | cap->arg);
+   cap->count1 -= 1;
+    }
 }

 /*
  * Search for "pat" in direction "dir" ('/' or '?', 0 for repeat).
  * Uses only cap->count1 and cap->oap from "cap".
- */
-    static void
+ * Return 0 for failure, 1 for found, 2 for found and line offset added.
+ */
+    static int
 normal_search(cap, dir, pat, opt)
     cmdarg_T   *cap;
     int        dir;
@@ -6354,6 +6366,7 @@ normal_search(cap, dir, pat, opt)
     /* "/$" will put the cursor after the end of the line, may need to
      * correct that here */
     check_cursor();
+    return i;
 }

 /*
diff --git a/src/nvim/testdir/test44.in b/src/nvim/testdir/test44.in
--- a/src/nvim/testdir/test44.in
+++ b/src/nvim/testdir/test44.in
@@ -42,6 +42,12 @@ G:put =matchstr(\"אבגד\", \".\", 0, 
 :put =matchstr(\"אבגד\", \"..\", 0, 2) " בג
 :put =matchstr(\"אבגד\", \".\", 0, 0) " א
 :put =matchstr(\"אבגד\", \".\", 4, -1) " ג
+:new
+:$put =['dog(a', 'cat(']
+/(/e+
+"ayn:bd!
+:$put =''
+G"ap
 :w!
 :qa!
 ENDTEST
diff --git a/src/nvim/testdir/test44.ok b/src/nvim/testdir/test44.ok
--- a/src/nvim/testdir/test44.ok
+++ b/src/nvim/testdir/test44.ok
@@ -22,3 +22,5 @@ k œ̄ṣ́m̥̄ᾱ̆́
 בג
 א
 ג
+a
+cat(
diff --git a/src/nvim/version.c b/src/nvim/version.c
--- a/src/nvim/version.c
+++ b/src/nvim/version.c
@@ -742,6 +742,8 @@ static char *(features[]) =
 static int included_patches[] =
 {   /* Add new patch number below this line */
 /**/
+    636,
+/**/
     635,
 /**/
     634,
```

### CODE REVIEW

File: src/nvim/normal.c [hunk 1]
```diff
@@ -4361,7 +4361,7 @@ static void nv_ident(cmdarg_T *cap)
     /* put pattern in search history */
     init_history();
     add_to_history(HIST_SEARCH, buf, true, NUL);
-    normal_search(cap, cmdchar == '*' ? '/' : '?', buf, 0);
+    (void)normal_search(cap, cmdchar == '*' ? '/' : '?', buf, 0);
```
Comment (glts) [inline src/nvim/normal.c]: Reviewers: is this still done nowadays, casting an unused return value to `void`? I haven't been writing any C lately.

Comment (jszakmeister) [inline src/nvim/normal.c]: Yes, if the intention is really to ignore the return value.

Comment (glts) [inline src/nvim/normal.c]: Ok, thanks.

**Comment by glts** [comment]
Unrelated test failure, triggering again.

**Comment by glts** [comment]
Rebased.

This is a small patch, should be safe. If somebody could give it a quick review (and merge :dog:?) I would immediately start porting legacy tests 44 and 99, the first of which I'm bringing up to date here.

**Comment by Pyrohh** [comment]
LGTM

**Comment by justinmk** [comment]
Merged, thanks!

**Pull request closed without merging**